In [1]:
import sklearn
print(sklearn.__version__)

# 사이킷런: 머신러닝 package로 가장 파이썬스로운 API를 제공

0.21.2


In [2]:
from sklearn.datasets import load_iris
# 사이킷런의 많은 샘플 데이터 중 붓꽃 품종 데이터를 가져옴

from sklearn.tree import DecisionTreeClassifier

from sklearn.model_selection import train_test_split
# 데이터를 train과 test로 분류해주는 라이브러리

import pandas as pd


In [3]:
# 우선 필요한 데이터를 불러옵니다

iris= load_iris()

iris_data = iris.data

iris_label = iris.target

print('iris target값: ', iris_label)
print('iris target명: ', iris.target_names)

# 이미 정제된 데이터이기에 iris.target_names 등를 뽑아낼 수 있다.
print('iris feature명: ', iris.feature_names)

iris target값:  [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]
iris target명:  ['setosa' 'versicolor' 'virginica']
iris feature명:  ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


In [4]:
iris_df = pd.DataFrame(data =iris_data, columns = iris.feature_names)
# 기존의 ndarray형태였떤 iris_data를 데이터프레임화 합니다.
# column은 feature_names로 세팅

iris_df['label'] = iris.target
iris_df.head()
# 서로 다른 데이터 꽃들이 150개 존재 

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   label  
0      0  
1      0  
2      0  
3      0  
4      0

# 데이터 살펴보기

In [5]:
iris_data = load_iris()
keys = iris_data.keys()
print('붓꽃 데이터 세트의 키들:', keys)

붓꽃 데이터 세트의 키들: dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])


In [6]:
print('\n feature_names 의 type:',type(iris_data.feature_names))
print(' feature_names 의 shape:',len(iris_data.feature_names))
print(iris_data.feature_names)

print('\n target_names 의 type:',type(iris_data.target_names))
print(' feature_names 의 shape:',len(iris_data.target_names))
print(iris_data.target_names)

print('\n data 의 type:',type(iris_data.data))
print(' data 의 shape:',iris_data.data.shape)
print(iris_data['data'])

print('\n target 의 type:',type(iris_data.target))
print(' target 의 shape:',iris_data.target.shape)
print(iris_data.target)


 feature_names 의 type: <class 'list'>
 feature_names 의 shape: 4
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']

 target_names 의 type: <class 'numpy.ndarray'>
 feature_names 의 shape: 3
['setosa' 'versicolor' 'virginica']

 data 의 type: <class 'numpy.ndarray'>
 data 의 shape: (150, 4)
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.2]
 [5.  3.2 1.2 0

## 학습/테스트 데이터 셋 분리 - train_test_split()

In [14]:
iris = load_iris() # iris data set 뽑아왔다.
dt_clf = DecisionTreeClassifier() # 디시션트리분류메서드 할당

train_data = iris.data # iris data에서 레이블이 아닌 부분(iris.data)
train_label = iris.target #iris data에서 label부분(iris.target)
#즉, data를 활용/추론하여 target이라는 결론을 이끌어낸다.(target:분류된 것)

dt_clf.fit(train_data, train_label) # 학습 진행
# train,test로 분류하지 않은 상태에서 학습 진행


# 학습 데이터 셋으로 예측 수행
pred = dt_clf.predict(train_data) 
#디시션트리분류 메서드의 predict 메서드로 train_data 인수로 받는다, pred에 저장.
print('예측 정확도: ', accuracy_score(train_label, pred))

예측 정확도:  1.0


###  train과 test로 분리 없이 학습을 진행하였기에 예측 정확도는 1이 나온다. (=예측을 모두 맞췄다는 뜻일 수도 있지만, 이미 정해져 있는 결과를 그대로 가져왔다는 뜻이기도 하다)

In [15]:
from sklearn.model_selection import train_test_split
iris_data = load_iris()
X_train, X_test, y_train, y_test = train_test_split(iris_data.data, iris_data.target, test_size = 0.3, random_state = 121)
# X_train, X_test는 학습용 feature 데이터 셋
# y_train, y_test는 학습용 레이블 데이터와 테스트용 레이블 데이터

dt_clf.fit(X_train, y_train) 
# DecisionTreeClasifier의 객체 dt_clf로 학습 데이터를 학습(fit)
# dt_clf객체는 학습한 데이터의 내용(?)을 기반으로, X_test 데이터를 예측(predict())한다.
pred = dt_clf.predict(X_test)

# 학습데이터로 학습한 알고리즘을 기반으로 새로운 데이터를 학습할 때, 얼마나 잘 예측했는지 검증
print('예측 정확도: {0: .4}'.format(accuracy_score(y_test, pred)))

예측 정확도:  0.9556


# Q. DecisionTreeClassifier는 어떤 알고리즘이길래??

.

In [21]:
import pandas as pd

iris_df = pd.DataFrame(iris_data.data, columns=iris_data.feature_names)
iris_df['target'] = iris_data.target
# iris_df['target']=iris_data.target

tr_df= iris_df.iloc[:,:-1]
tgt_df = iris_df.iloc[:,-1]
X_train, X_test, y_train, y_test = train_test_split(tr_df, tgt_df, test_size = 0.3, random_state=121)

In [22]:
print(type(X_train),type(X_test),type(y_train),type(y_test))

<class 'pandas.core.frame.DataFrame'> <class 'pandas.core.frame.DataFrame'> <class 'pandas.core.series.Series'> <class 'pandas.core.series.Series'>


In [23]:
dt_clf.fit(X_train, y_train)
pred = dt_clf.predict(X_test)
print("예측 정확도: {0: .4f}".format(accuracy_score(y_test,pred)))

예측 정확도:  0.9556


# 교차 검증

## K 폴드

In [24]:
from sklearn.model_selection import KFold
import numpy as np

iris = load_iris() # 데이터 가져오기
features = iris.data
label = iris.target
dt_clf = DecisionTreeClassifier(random_state=156) # features, label, DecisionTreeClassifer 호출

# 5개의 폴드 세트로 분리하는 KFold 객체 (KFold 라이브러리에서 kfold객체 선언)
kfold = KFold(n_splits=5) 

# 폴드 세트별 정확도를 담을 리스트 객체 생성
cv_accuracy = []
print('붓꽃 데이터 세트 크기:',features.shape[0])

# row가 150개라는 뜻이다

붓꽃 데이터 세트 크기: 150


In [25]:
n_iter = 0

# KFold객체의 split( ) 호출하면 
# 폴드 별 학습용, 검증용 테스트의 로우 인덱스를 array로 반환  
for train_index, test_index  in kfold.split(features):
    
    # kfold.split( )으로 반환된 인덱스를 이용하여 학습용 데이터 추출
    X_train, X_test = features[train_index], features[test_index]
    
    # 검증용 테스트 데이터 추출
    y_train, y_test = label[train_index], label[test_index]
    
    #학습 및 예측 
    dt_clf.fit(X_train , y_train) 
    # 이전에는 그냥 train_test_split()으로 나눴는데, 이번엔 KFold로 train과 test분류함
    
    # fit --> predict
    pred = dt_clf.predict(X_test) # 즉 pred는 y_test를 예측한 것
    n_iter += 1
    
    # 반복 시 마다 정확도 측정 
    accuracy = np.round(accuracy_score(y_test,pred), 4)
    train_size = X_train.shape[0]
    test_size = X_test.shape[0]
    print('\n#{0} 교차 검증 정확도 :{1}, 학습 데이터 크기: {2}, 검증 데이터 크기: {3}'
          .format(n_iter, accuracy, train_size, test_size))
    print('#{0} 검증 세트 인덱스:{1}'.format(n_iter,test_index))
    
    cv_accuracy.append(accuracy)
    
# 개별 iteration별 정확도를 합하여 평균 정확도 계산 
print('\n## 평균 검증 정확도:', np.mean(cv_accuracy))


#1 교차 검증 정확도 :1.0, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#1 검증 세트 인덱스:[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29]

#2 교차 검증 정확도 :0.9667, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#2 검증 세트 인덱스:[30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53
 54 55 56 57 58 59]

#3 교차 검증 정확도 :0.8667, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#3 검증 세트 인덱스:[60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83
 84 85 86 87 88 89]

#4 교차 검증 정확도 :0.9333, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#4 검증 세트 인덱스:[ 90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119]

#5 교차 검증 정확도 :0.7333, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#5 검증 세트 인덱스:[120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137
 138 139 140 141 142 143 144 145 146 147 148 149]

## 평균 검증 정확도: 0.9


# 즉, train_test_split()은 학습-검증 데이터 세트로 한번만 나누는 반면, 
# 교차검증(KFold)는 서로 다른 학습-검증 데이터 세트를 여러개 만들어 
# DecisionTreeClassifier로 학습시킨다

###  동일한 데이터 세트를 어떻게 나누느냐의 차이

.

# Stratified K폴드

- 기존 k 폴드

In [26]:
iris = load_iris()
iris_df = pd.DataFrame(data=iris.data, columns = iris.feature_names)
iris_df['label']=iris.target
iris_df['label'].value_counts()  
# valuecounts = 해당 데이터가 몇개나 있는지 분류해주는 메소드

# 전체 150개의 데이터가 3개의 분류로 나뉘어 있다

2    50
1    50
0    50
Name: label, dtype: int64

In [27]:
kfold = KFold(n_splits = 3)

n_iter = 0

# train_test_split() => kfold.split()
for train_index, test_index in kfold.split(iris_df):
    n_iter +=1
    label_train = iris_df['label'].iloc[train_index]
    label_test = iris_df['label'].iloc[test_index]
    
    print('\n## 교차 검증: {0}'.format(n_iter))
    print('train label 데이터 분포: \n',label_train.value_counts())
    print('test label 데이터 분포: \n',label_test.value_counts())


## 교차 검증: 1
train label 데이터 분포: 
 2    50
1    50
Name: label, dtype: int64
test label 데이터 분포: 
 0    50
Name: label, dtype: int64

## 교차 검증: 2
train label 데이터 분포: 
 2    50
0    50
Name: label, dtype: int64
test label 데이터 분포: 
 1    50
Name: label, dtype: int64

## 교차 검증: 3
train label 데이터 분포: 
 1    50
0    50
Name: label, dtype: int64
test label 데이터 분포: 
 2    50
Name: label, dtype: int64


- Stratified k폴드

In [28]:
from sklearn.model_selection import StratifiedKFold

skf = StratifiedKFold(n_splits = 3)
n_iter = 0

#위의 과정과 비교해서 다른 점은, 이 코드는 전체 데이터와 레이블 데이터로 쪼갠다.
# 위의 코드가 한 세션당 3개로 쪼개졌다면, 여긴 6개로 쪼개진다.

for train_index, test_index in skf.split(iris_df, iris_df['label']):
    n_iter += 1
    label_train= iris_df['label'].iloc[train_index]
    label_test= iris_df['label'].iloc[test_index]
    print('\n## 교차 검증: {0}'.format(n_iter))
    print('train label 데이터 분포:\n', label_train.value_counts(), '\n')
    print('test label 데이터 분포:\n', label_test.value_counts())
    print('---------------------')

# 그래서 이건, 이전에 학습데이터 뭉텅이, 그 안에서 더 쪼개 검증을 하겠단 소리여



## 교차 검증: 1
train label 데이터 분포:
 2    33
1    33
0    33
Name: label, dtype: int64 

test label 데이터 분포:
 2    17
1    17
0    17
Name: label, dtype: int64
---------------------

## 교차 검증: 2
train label 데이터 분포:
 2    33
1    33
0    33
Name: label, dtype: int64 

test label 데이터 분포:
 2    17
1    17
0    17
Name: label, dtype: int64
---------------------

## 교차 검증: 3
train label 데이터 분포:
 2    34
1    34
0    34
Name: label, dtype: int64 

test label 데이터 분포:
 2    16
1    16
0    16
Name: label, dtype: int64
---------------------


# StratifiedKFold, not KFold

In [29]:
skFold = StratifiedKFold(n_splits = 3)
n_iter = 0
cv_accuracy = []

# features와 label은 어디서 정의 된 것?
for train_index, test_index in skFold.split(features, label):
    X_train, X_test = features[train_index], features[test_index]
    y_train, y_test = label[train_index], label[test_index]
    
    dt_clf.fit(X_train, y_train)
    pred = dt_clf.predict(X_test)
    
    n_iter += 1
    accuracy = np.round(accuracy_score(y_test, pred),3)
    train_size = X_train.shape[0]
    test_size = X_test.shape[0]
    
    print('\n#{0} 교차 검증 정확도 :{1}, 학습 데이터 크기: {2}, 검증 데이터 크기: {3}'
          .format(n_iter, accuracy, train_size, test_size))
    print('#{0} 검증 세트 인덱스:{1}'.format(n_iter,test_index))
    cv_accuracy.append(accuracy)
print("\n ## 교차 검증별 정확도:", np.round(cv_accuracy, 3))
print("## 평균 검증 정확도:", np.mean(cv_accuracy))

# 왜 마지막 검증에서 인덱스가 하나 더 추가 됐는지 알겠네. fucking logical




#1 교차 검증 정확도 :0.98, 학습 데이터 크기: 99, 검증 데이터 크기: 51
#1 검증 세트 인덱스:[  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  50
  51  52  53  54  55  56  57  58  59  60  61  62  63  64  65  66 100 101
 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116]

#2 교차 검증 정확도 :0.922, 학습 데이터 크기: 99, 검증 데이터 크기: 51
#2 검증 세트 인덱스:[ 17  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  67
  68  69  70  71  72  73  74  75  76  77  78  79  80  81  82  83 117 118
 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133]

#3 교차 검증 정확도 :0.979, 학습 데이터 크기: 102, 검증 데이터 크기: 48
#3 검증 세트 인덱스:[ 34  35  36  37  38  39  40  41  42  43  44  45  46  47  48  49  84  85
  86  87  88  89  90  91  92  93  94  95  96  97  98  99 134 135 136 137
 138 139 140 141 142 143 144 145 146 147 148 149]

 ## 교차 검증별 정확도: [0.98  0.922 0.979]
## 평균 검증 정확도: 0.9603333333333334


- 모든 분류 방식은 DecisionTreeClassifier를 기반으로 한다

- train_test_split: 0.9556
- KFold: 0.9
- StratifiedKFold: 0.96
- 결국 StratifiedKFold의 성능이 KFold보다 조금 더 낫다는 걸 확인



# cross_val_score()

- 이전의 방법들 보다 간단할 뿐더러 성능도 좋다

In [31]:
from sklearn.model_selection import cross_val_score , cross_validate
from sklearn.datasets import load_iris
import numpy as np

iris_data =load_iris()
dt_clf = DecisionTreeClassifier(random_state = 156)

data = iris_data.data
label = iris_data.target

scores = cross_val_score(dt_clf, data, label, scoring = 'accuracy', cv = 3)
# cross_val_score : 교차검증 메소드(디시션트리, 피쳐, 레이블, 정확도와 분류 세개..?)
print('교차 검증별 정확도: ',np.round(scores, 4))
print('평균 검증 정확도: ', np.round(np.mean(scores),4))


교차 검증별 정확도:  [0.9804 0.9216 0.9792]
평균 검증 정확도:  0.9604


# GridSearchCV는 따로 정리..이해를 못했거든요

In [ ]:
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.metrics import accuracy_score

X_train, X_test, y_train, y_test = train_test_split(iris_data.data, iris_data.target, test_size = 0.2, random_state = 121)
dtree = DecisionTreeClassifier(random_state = 126)

### parameter 들은 dictionary 형태로 설정
parameters = {'max_depth':[1,2,3], 'min_samples_split':[2,3]}

In [ ]:
import pandas as pd

grid_dtree = GridSearchCV(dtree, param_grid = parameters, cv = 3, refit = True, return_train_score=True)

# 붓꽃 Train 데이터로 param_grid의 하이퍼 파라미터들을 순차적으로 학습/평가 .
grid_dtree.fit(X_train, y_train)

# GridSearchCV 결과는 cv_results_ 라는 딕셔너리로 저장됨. 이를 DataFrame으로 변환
scores_df = pd.DataFrame(grid_dtree.cv_results_)
scores_df[['params', 'mean_test_score', 'rank_test_score', 'split0_test_score', 'split1_test_score', 'split2_test_score']]